dlt/destinations/sql_jobs.py

```diff
@@ -515,6 +550,84 @@ def gen_merge_sql(
             sql.append(f"INSERT INTO {table_name}({col_str}) {select_sql};")
         return sql
 
+    @classmethod
+    def gen_upsert_sql(
+        cls, table_chain: Sequence[TTableSchema], sql_client: SqlClientBase[Any]
+    ) -> List[str]:
+        sql: List[str] = []
+        root_table = table_chain[0]
+        root_table_name, staging_root_table_name = sql_client.get_qualified_table_names(
+            root_table["name"]
+        )
+        escape_id, escape_lit = sql_client.capabilities.escapers
+
+        # process table hints
+        primary_keys = cls._escape_list(
+            get_columns_names_with_prop(root_table, "primary_key"),
+            escape_id,
+        )
+        hard_delete_col, deleted_cond = cls._get_hard_delete_col_and_cond(
+            root_table,
+            escape_id,
+            escape_lit,
+        )
+
+        # generate merge statement for root table
+        on_str = " AND ".join([f"d.{c} = s.{c}" for c in primary_keys])
+        root_table_column_names = list(map(escape_id, root_table["columns"].keys()))
+        update_str = ", ".join([c + " = " + "s." + c for c in root_table_column_names])
+        col_str = ", ".join(["{alias}" + c for c in root_table_column_names])
+        delete_str = (
+            "" if hard_delete_col is None else f"WHEN MATCHED AND s.{deleted_cond} THEN DELETE"
+        )
+
+        sql.append(f"""
+            MERGE INTO {root_table_name} d USING {staging_root_table_name} s
+            ON {on_str}
+            {delete_str}
+            WHEN MATCHED
+                THEN UPDATE SET {update_str}
+            WHEN NOT MATCHED
+                THEN INSERT ({col_str.format(alias="")}) VALUES ({col_str.format(alias="s.")});
+        """)
+
+        # generate statements for child tables if they exist
+        child_tables = table_chain[1:]
+        if child_tables:
+            root_row_key = escape_id("_dlt_id")
```

Comment:
Should the row id be marked with a new hint (other than `unique`) to prevent hard-coding `_dlt_id`?

Reply:
why not `unique`? technically you just need a unique column to delete / merge child tables correctly. `delete-insert` does that, right?

Reply:
no, we do not have such test case. what I want to achieve here is that we rely on annotations, not on column names.
- `_dlt_id` is annotated as `unique` by a standard schema
- `root_key` is a also an annotation, not a hardcoded field
`relational.py` adds config for each table with `merge` write disposition:
```
"propagation": {
                        "tables": {
                            table_name: {
                                TColumnName(self.c_dlt_id): TColumnName(self.c_dlt_root_id)
                            }
                        }
                    }
```
which will propagate `_dlt_id` from root to each child table as `_dlt_root_id` and
```
self.schema._merge_hints(
            {
                "not_null": [
                    TSimpleRegex(self.c_dlt_id),
                    TSimpleRegex(self.c_dlt_root_id),
                    TSimpleRegex(self.c_dlt_parent_id),
                    TSimpleRegex(self.c_dlt_list_idx),
                    TSimpleRegex(self.c_dlt_load_id),
                ],
                "foreign_key": [TSimpleRegex(self.c_dlt_parent_id)],
                "root_key": [TSimpleRegex(self.c_dlt_root_id)],
                "unique": [TSimpleRegex(self.c_dlt_id)],
            },
            normalize_identifiers=False,  # already normalized
        )
```
makes sure that hints are applied. this is quite old version of `dlt` and we'll reimplement it. but high level mechanism (we rely on annotations, not names) is good